lib/ipam/ipam.go

```diff
@@ -1414,24 +1414,29 @@ func (c ipamClient) SetIPAMConfig(ctx context.Context, cfg IPAMConfig) error {
 		return errors.New("Cannot disable 'StrictAffinity' and 'AutoAllocateBlocks' at the same time")
 	}
 
-	allObjs, err := c.client.List(ctx, model.BlockListOptions{}, "")
+	// Get revision if resource already exists
+	old, err := c.client.Get(ctx, model.IPAMConfigKey{}, "")
 	if err != nil {
-		return err
-	}
-	if len(allObjs.KVPairs) != 0 {
-		return errors.New("Cannot change IPAM config while allocations exist")
+		if _, ok := err.(cerrors.ErrorResourceDoesNotExist); !ok {
+			log.Errorf("Error querying IPAMConfig %v", err)
+			return err
+		}
 	}
 
 	// Write to datastore.
 	obj := model.KVPair{
 		Key:   model.IPAMConfigKey{},
 		Value: c.convertIPAMConfigToBackend(&cfg),
 	}
+	if old != nil {
```

Comment:
I think we want to enforce that the calling code keeps the revision up-to-date, rather than always update it here.

That way we can get proper compare-and-swap behavior out of this function.

e.g., if two clients call `SetIPAMConfig` at the same time, only one will win and the other will need to retry.

Reply:
Hm, I see now, this function is a bit odd. We query it up front and compare to see if there's nothing to do.

I think we _could_ change this to not need to query the old state, and instead it just does an Apply on the provided object, but that is a change from the current behavior. It does, however, make it more aligned with how other create / update functions behave. 

So, I guess you're good to leave this as-is for now.